pkg/plugins/localfiles.go

```diff
@@ -89,10 +162,83 @@ func (f LocalFS) Remove() error {
 			return ErrUninstallInvalidPluginDir
 		}
 	}
-
 	return os.RemoveAll(f.basePath)
 }
 
+// staticFilesMap is a set-like map that contains files that can be accessed from a plugins.FS.
+type staticFilesMap map[string]struct{}
+
+// isAllowed returns true if the provided path is allowed.
+// path is a string accepted by an FS Open() method.
+func (a staticFilesMap) isAllowed(path string) bool {
+	_, ok := a[path]
+	return ok
+}
+
+// newStaticFilesMap creates a new staticFilesMap from a list of allowed file paths.
+func newStaticFilesMap(files ...string) staticFilesMap {
+	m := staticFilesMap(make(map[string]struct{}, len(files)))
+	for _, k := range files {
+		m[k] = struct{}{}
+	}
+	return m
+}
+
+// StaticFS wraps an FS and allows accessing only the files in the allowList.
+// This is a more secure implementation of a FS suitable for production environments.
+// The keys of the allow list must be in the same format used by the underlying FS' Open() method.
+type StaticFS struct {
+	FS
+
+	// staticFilesMap is a map of allowed paths (accepted by FS.Open())
+	staticFilesMap staticFilesMap
+}
+
+// NewStaticFS returns a new StaticFS that can access the files on an underlying FS,
+// but only if they are also specified in a static list, which is constructed when creating the object
+// by calling Files() on the underlying FS.
+func NewStaticFS(fs FS) (StaticFS, error) {
+	files, err := fs.Files()
+	if err != nil {
+		return StaticFS{}, err
+	}
+	return StaticFS{
+		FS:             fs,
+		staticFilesMap: newStaticFilesMap(files...),
+	}, nil
+}
+
+// Open checks that name is an allowed file and, if so, it returns a fs.File to access it, by calling the
+// underlying FS' Open() method.
+// If access is denied, the function returns ErrFileNotExist.
+func (f StaticFS) Open(name string) (fs.File, error) {
+	// Ensure access to the file is allowed
+	if !f.staticFilesMap.isAllowed(name) {
+		return nil, ErrFileNotExist
+	}
+	// Use the wrapped FS to access the file
+	return f.FS.Open(name)
+}
+
+// Files returns a slice of all the file paths in the FS relative to the base path.
+// It calls Files() on the underlying FS, and intersects the result with the allow-list.
+func (f StaticFS) Files() ([]string, error) {
+	// Get files from the underlying FS
+	filesystemFiles, err := f.FS.Files()
```

Comment:
Couldn't we skip this and just return what's inside the `f.staticFilesMap`?

Reply:
Good point. You're right, this can be changed due to the other changes done as part of the PR review